src/docquery/ext/pipeline_document_classification.py

```diff
@@ -0,0 +1,333 @@
+                            bbox.append([1000] * 4)
+                        else:
+                            bbox.append([0] * 4)
+
+                    span_encoding["bbox"] = torch.tensor(bbox).unsqueeze(0)
+
+                    yield {
+                        **span_encoding,
+                        "page": page_idx,
+                    }
+
+    def _forward(self, model_inputs):
+        page = model_inputs.pop("page", None)
+
+        if "overflow_to_sample_mapping" in model_inputs:
+            model_inputs.pop("overflow_to_sample_mapping")
+
+        if self.model_type == ModelType.VisionEncoderDecoder:
+            model_outputs = self.model.generate(**model_inputs)
+        else:
+            model_outputs = self.model(**model_inputs)
+
+        model_outputs["page"] = page
+        model_outputs["attention_mask"] = model_inputs.get("attention_mask", None)
+        return model_outputs
+
+    def postprocess(self, model_outputs, function_to_apply=None, top_k=1, **kwargs):
+        if function_to_apply is None:
+            if self.model.config.num_labels == 1:
+                function_to_apply = ClassificationFunction.SIGMOID
+            elif self.model.config.num_labels > 1:
+                function_to_apply = ClassificationFunction.SOFTMAX
+            elif hasattr(self.model.config, "function_to_apply") and function_to_apply is None:
+                function_to_apply = self.model.config.function_to_apply
+            else:
+                function_to_apply = ClassificationFunction.NONE
+
+        if self.model_type == ModelType.VisionEncoderDecoder:
+            answers = self.postprocess_encoder_decoder(model_outputs, top_k=top_k, **kwargs)
+        else:
+            answers = self.postprocess_standard(
+                model_outputs, function_to_apply=function_to_apply, top_k=top_k, **kwargs
+            )
+
+        answers = sorted(answers, key=lambda x: x.get("score", 0), reverse=True)[:top_k]
+        return answers
+
+    def postprocess_encoder_decoder(self, model_outputs, **kwargs):
+        classes = set()
+        for model_output in model_outputs:
+            for sequence in self.tokenizer.batch_decode(model_output.sequences):
+                sequence = sequence.replace(self.tokenizer.eos_token, "").replace(self.tokenizer.pad_token, "")
+                sequence = re.sub(r"<.*?>", "", sequence, count=1).strip()  # remove first task start token
+                classes.add(donut_token2json(self.tokenizer, sequence)["class"])
+
+        # Return the first top_k unique classes we see
+        return [{"label": v} for v in classes]
+
+    def postprocess_standard(self, model_outputs, function_to_apply, **kwargs):
+        # Average the score across pages
+        sum_scores = {k: 0 for k in self.model.config.id2label.values()}
+        for model_output in model_outputs:
+            outputs = model_output["logits"][0]
+            outputs = outputs.numpy()
+
```

Comment:
If we want to extend the model to be able to predict "No value", this can be done pretty easily by adding an additional mask here `outputs > 0` (this corresponds to a sigmoid output of > 0.5, the typical threshold for binary classification). In other words, the results returned by `postprocess_standard()` would _only_ include the top_k predictions in which their respective logit was > 0.

I'm not sure how to extend this to account for when `len(model_outputs) > 1`, though. Something I'll have to think about for a bit

Reply:
Okay makes sense. Maybe you could add this as a follow up PR? Filed here so we don't lose it: https://github.com/impira/docquery/issues/38

Reply:
👍 Sounds good, happy to lead this change